packages/tabular/src/Happy/Tabular.hs

```diff
@@ -0,0 +1,123 @@
+  infoFile :: Maybe String
+  
+#ifdef DEBUG
+
+  , 
+  dumpLR0 :: Bool,
+  dumpLA :: Bool,
+  dumpAction :: Bool,
+  dumpGoto :: Bool
+
+#endif
+
+}
+
+type Lr1State = ([Lr1Item], [(Name, Int)])
+type TabularResult = (ActionTable, GotoTable, [Lr1State], [Int])
+
+runTabular :: TabularArgs -> Bool -> Grammar -> IO TabularResult
+runTabular args glr g = 
+    let first       = {-# SCC "First" #-} (mkFirst g)
+        closures    = {-# SCC "Closures" #-} (precalcClosure0 g)
+        sets        = {-# SCC "LR0_Sets" #-} (genLR0items g closures)
+        _lainfo@(spont,prop) = {-# SCC "Prop" #-} (propLookaheads g sets first)
+        la          = {-# SCC "Calc" #-} (calcLookaheads (length sets) spont prop)
+        items2      = {-# SCC "Merge" #-} (mergeLookaheadInfo la sets)
+        goto        = {-# SCC "Goto" #-} (genGotoTable g sets)
+        action      = {-# SCC "Action" #-} (genActionTable g first items2)
+        (conflictArray,(sr,rr))   = {-# SCC "Conflict" #-} (countConflicts action)
+    in do
+
+#ifdef DEBUG
+
+        optPrint (dumpLR0 args) (print sets)
+        optPrint (dumpLA args) (print la)
+        optPrint (dumpAction args) (print action)
+        optPrint (dumpGoto args) (print goto)
+
+#endif
+
+        (unused_rules, unused_terminals) <- reportUnusedRules glr g action
+        writeInfoFile sets g action goto conflictArray (inFile args) (infoFile args) unused_rules unused_terminals
+        reportConflicts g sr rr
+        return (action, goto, items2, unused_rules)
+
+#ifdef DEBUG
+
+    where
+      optPrint b io = when b (putStr "\n---------------------\n" >> io)
+
+#endif
+
+
+-------- Helpers --------
+
+reportUnusedRules :: Bool -> Grammar -> ActionTable -> IO ([Int], [String])
+reportUnusedRules glr g action = 
+    let reduction_filter = if glr then any_reduction else first_reduction
+        (unused_rules, unused_terminals) = find_redundancies reduction_filter g action in do
+    when (not (null unused_rules))
+       (hPutStrLn stderr ("unused rules: " ++ show (length unused_rules)))
+    when (not (null unused_terminals))
+       (hPutStrLn stderr ("unused terminals: " ++ show (length unused_terminals)))
+    return (unused_rules, unused_terminals)
+
+reportConflicts :: Grammar -> Int -> Int -> IO ()
+reportConflicts g sr rr = case expect g of
+    Just n | n == sr && rr == 0 -> return ()
+    Just _ | rr > 0 ->
+        die' $ "The grammar has reduce/reduce conflicts.\n" ++
+               "This is not allowed when an expect directive is given\n"
+    Just _ ->
+        die' $ "The grammar has " ++ show sr ++ " shift/reduce conflicts.\n" ++
+               "This is different from the number given in the expect directive\n"
+    _ -> do    
+        if sr /= 0
+          then hPutStrLn stderr ("shift/reduce conflicts:  " ++ show sr)
+          else return ()
+        
+        if rr /= 0
+          then hPutStrLn stderr ("reduce/reduce conflicts: " ++ show rr)
+          else return ()
+
+die' :: String -> IO a
```

Comment:
Why is this named `die'` with a prime? Is this somehow different from `die`?

Reply:
No, `die` is exported by `System.Exit`, but only starting from base-4.8.0.0. For calling it `die` we'd need an `#if MIN_VERSION_base`, which will come in the following MR.

Reply:
Hm, what about using an explicit import?

```
import System.Exit (exitWith, ExitCode(..))
```

Reply:
True, this would work. The next MR will remove `die'` anyhow and replace it with a local function named `die`, so it doesn‘t really make any difference.